tket/src/WeightSubgrMono/Searching/ValueOrdering.cpp

```diff
@@ -0,0 +1,117 @@
+// Copyright 2019-2022 Cambridge Quantum Computing
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "WeightSubgrMono/Searching/ValueOrdering.hpp"
+
+#include <algorithm>
+
+#include "Utils/Assert.hpp"
+#include "Utils/RNG.hpp"
+#include "WeightSubgrMono/GraphTheoretic/NeighboursData.hpp"
+
+namespace tket {
+namespace WeightedSubgraphMonomorphism {
+
+/*
+We are using "solution biased searching" as our heuristic. See the paper
+"Sequential and Parallel Solution-Biased Search for Subgraph Algorithms".
+The idea is, rather than always mapping into the target vertex with largest
+possible degree, we should try lower degrees occasionally.
+This is done simply by choosing a random vertex,
+but letting the probability of choosing a vertex depend on the degree,
+so that lower degrees are less likely.
+*/
+
+ValueOrdering::ValueOrdering() {
+  // 2^{-5} = 1/32 ~ 0.03. A choice with probability <3% is probably
+  // not very significant for the overall search.
+  m_data.resize(5);
```

Comment:
No. 5 refers to degrees, not vertices. The probability of picking a vertex (a mass) is proportional to 2^(-degree). Therefore if a vertex has <= max degree - 5, then it has weighting <= 1/32 ~ 3% of any vertex of maximum degree.

Thus, including these vertices also would only make a difference in a very small number of cases.

The point (as explained in the linked paper) is that our heuristic is to choose highest degree vertices first, but occasionally we choose slightly lower degrees simply because of a very rough estimate that the expected number of solutions in the subtree is roughly C . 2^degree.

However, this is all very rough and experimental (as the paper says); it gives better results than non-exponential formulae but there's no deep underlying theory here.

The selection "5" is a further arbitrary choice; we need a cutoff to avoid wasting time on vanishingly unlikely choices with this heuristic, and 3% seems small enough that it's a reasonable cutoff. (OK, we could have 1 vertex of degree 10 and 1000 vertices of degree 5. But even in that case, I think we'd rather just try the degree 10 vertex first).

Reply:
Thanks a lot. Please could you add this explanation to the code - I appreciate its in the paper but it seems important enough that its worth making abundantly clear to the code peruser.

Reply:
Done, in next push! (Actually it's proportional to 2^degree, we prefer HIGHER degrees).